typhos/widgets.py

```diff
@@ -256,16 +277,34 @@ def unit_changed(self, new_unit):
         new_unit : str
             The new unit
         """
-        if self._unit != new_unit:
-            super().unit_changed(new_unit)
-            if new_unit.lower() in EXPONENTIAL_UNITS \
-                    and self.displayFormat == PyDMLabel.DisplayFormat.Default:
-                self.displayFormat = PyDMLabel.DisplayFormat.Exponential
-
+        if self._unit == new_unit:
+            return
 
-class TyphosLabel(PyDMLabel):
+        super().unit_changed(new_unit)
+        default = (self.displayFormat == DisplayFormat.Default)
+        if new_unit.lower() in EXPONENTIAL_UNITS and default:
+            self.displayFormat = DisplayFormat.Exponential
+
+
+@use_for_variety_read('array-nd')
+@use_for_variety_read('command')
+@use_for_variety_read('command-enum')
+@use_for_variety_read('command-proc')
```

Comment:
I think the decorator stack is a pretty good solution to categorizing/registering these widgets

Reply:
Thanks! (and phew, I was a bit worried I got _too_ carried away here with magic...)

Reply:
I think we have worse offenders lying around when it comes to magic... Magic is only a real problem when it makes the code hard to read. I think decorators like these make it easier to read.